File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/library/items/media/files/cached/stream/specs/GivenATypicalFile/WhenCommittingTheFileToCache.java
package com.lasthopesoftware.bluewater.client.browsing.library.items.media.files.cached.stream.specs.GivenATypicalFile;

import com.lasthopesoftware.bluewater.client.browsing.library.items.media.files.cached.stream.CachedFileOutputStream;
import com.namehillsoftware.handoff.promises.Promise;

import org.junit.BeforeClass;
import org.junit.Test;

import java.io.File;
import java.util.concurrent.CountDownLatch;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;

public class WhenCommittingTheFileToCache {

	private static final File mockedFile = mock(File.class);
	private static File persistedFile;
	private static String persistedKey;

	@BeforeClass
	public static void before() throws InterruptedException {
		final CachedFileOutputStream cachedFileOutputStream = new CachedFileOutputStream("unique-test", mockedFile, (uniqueKey, file) -> {
			persistedKey = uniqueKey;
			persistedFile = file;
			return Promise.empty();
		});

		final CountDownLatch countDownLatch = new CountDownLatch(1);
		cachedFileOutputStream.commitToCache()
			.then(v -> {
				countDownLatch.countDown();
				return null;
			});

		countDownLatch.await();
	}

	@Test
	public void thenTheCorrectKeyIsPersisted() {
		assertThat(persistedKey).isEqualTo("unique-test");
	}

	@Test
	public void thenTheCorrectFileIsPersisted() {
		assertThat(persistedFile).isEqualTo(mockedFile);
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/browsing/files/cached/stream/GivenATypicalFile/WhenCommittingTheFileToCache.kt
package com.lasthopesoftware.bluewater.client.browsing.files.cached.stream.GivenATypicalFile

import com.lasthopesoftware.bluewater.client.browsing.files.cached.stream.CachedFileOutputStream
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.io.File

class WhenCommittingTheFileToCache {
	private val mockedFile = mockk<File>()
	private var persistedLibrary: LibraryId? = null
	private var persistedFile: File? = null
	private var persistedKey: String? = null

	@BeforeAll
	fun act() {
		val cachedFileOutputStream = CachedFileOutputStream(
			LibraryId(156),
			"unique-test",
			mockedFile,
			mockk {
				every { putIntoDatabase(any(), any(), any()) } answers {
					persistedLibrary = firstArg()
					persistedKey = secondArg()
					persistedFile = lastArg()
					Promise.empty()
				}
			})
		cachedFileOutputStream.commitToCache().toExpiringFuture().get()
	}

	@Test
	fun `then the correct library is persisted`() {
		assertThat(persistedLibrary).isEqualTo(LibraryId(156))
	}

	@Test
	fun thenTheCorrectKeyIsPersisted() {
		assertThat(persistedKey).isEqualTo("unique-test")
	}

	@Test
	fun thenTheCorrectFileIsPersisted() {
		assertThat(persistedFile).isEqualTo(mockedFile)
	}
}
